src/core/attributeformmodelbase.cpp

```diff
@@ -632,25 +642,28 @@ void AttributeFormModelBase::updateEditorWidgetCodes( const QString &fieldName )
 
     if ( item->data( AttributeFormModel::ElementType ) == QStringLiteral( "qml" ) || item->data( AttributeFormModel::ElementType ) == QStringLiteral( "html" ) )
     {
-      const thread_local QRegularExpression sRegEx( "expression\\.evaluate\\(\\s*\\\"(.*?[^\\\\])\\\"\\s*\\)", QRegularExpression::MultilineOption | QRegularExpression::DotMatchesEverythingOption );
-      QRegularExpressionMatchIterator matchIt = sRegEx.globalMatch( code );
-      while ( matchIt.hasNext() )
+      if ( !mEditorWidgetCodesRequirements.contains( code ) )
       {
-        const QRegularExpressionMatch match = matchIt.next();
-        QString expression = match.captured( 1 );
-        expression = expression.replace( QStringLiteral( "\\\"" ), QStringLiteral( "\"" ) );
-
-        QgsExpression exp( expression );
-        exp.prepare( &mExpressionContext );
-        if ( exp.referencedColumns().contains( fieldName ) || exp.referencedColumns().contains( QgsFeatureRequest::ALL_ATTRIBUTES ) )
+        CodeRequirements codeRequirements;
+        const thread_local QRegularExpression sRegEx( "expression\\.evaluate\\(\\s*\\\"(.*?[^\\\\])\\\"\\s*\\)", QRegularExpression::MultilineOption | QRegularExpression::DotMatchesEverythingOption );
+        QRegularExpressionMatchIterator matchIt = sRegEx.globalMatch( code );
+        while ( matchIt.hasNext() )
         {
-          needUpdate = true;
-          break;
+          const QRegularExpressionMatch match = matchIt.next();
+          QString expression = match.captured( 1 );
+          expression = expression.replace( QStringLiteral( "\\\"" ), QStringLiteral( "\"" ) );
+
+          QgsExpression exp( expression );
+          exp.prepare( &mExpressionContext );
+          codeRequirements.referencedColumns.unite( exp.referencedColumns() );
+          codeRequirements.formScope = codeRequirements.formScope || QgsValueRelationFieldFormatter::expressionRequiresFormScope( expression );
         }
+        mEditorWidgetCodesRequirements.insert( code, codeRequirements );
       }
 
-      if ( needUpdate )
+      if ( mEditorWidgetCodesRequirements[code].referencedColumns.contains( fieldName ) || mEditorWidgetCodesRequirements[code].referencedColumns.contains( QgsFeatureRequest::ALL_ATTRIBUTES ) || mEditorWidgetCodesRequirements[code].formScope )
```

Comment:
I think that's duplicated on line 719, I think it would make sense to deduplicate in a function with the additional benefit of being able to give the function a name that adds semantics.

Reply:
@m-kuhn , done.